source/sema.h

```diff
@@ -19,23 +19,147 @@
+    { }
+
+    void load(std::string_view lib) {
+        active_instance = this;
+        in_progress = lib;
+        finally _{ [&](){ active_instance = nullptr; in_progress = ""; } };
+        auto& dl = loaded.emplace_back(in_progress);
+        if(!dl.is_open()) {
+            errors.emplace_back(
+                source_position(-1, -1),
+                "unable to load library '" + in_progress + "': " + dll::get_last_error()
+            );
+        }
+    }
+
+    bool try_apply(
+        [[maybe_unused]] std::string_view caller,
+        std::string const& name,
+        auto& decl
+    ) const {
+        // TODO(DyXel): Do full/better name look-up.
+        auto search = functions.find(name);
+        if(search == functions.end()) return false;
+        auto& entry = search->second;
+        auto visitor = [&](auto&& v) -> bool {
+            if constexpr(requires { v(decl); }) {
+                v(decl);
+                return true;
+            }
+            return false;
+        };
+        if constexpr(std::is_same_v<meta::type_declaration, CPP2_TYPEOF(decl)>)
+            return std::visit(visitor, entry.type);
+        else
+            return std::visit(visitor, entry.func);
+    }
+
+private:
+    friend meta::register_metafunction;
+
+    void register_one(std::string name, auto* func) {
+        // TODO(DyXel): Check if already registered
+        auto& entry = functions[name];
+        if constexpr(requires { entry.type = func; })
+            entry.type = func;
+        else
+            entry.func = func;
+    }
+};
+
+namespace meta {
+
+register_metafunction::register_metafunction(const char* name, mf_sign_type_in* f) {
+    active_instance->register_one(name, f);
+}
+
+register_metafunction::register_metafunction(const char* name, mf_sign_type_inout* f) {
+    active_instance->register_one(name, f);
+}
+
+register_metafunction::register_metafunction(const char* name, mf_sign_func_in* f) {
+    active_instance->register_one(name, f);
+}
+
+register_metafunction::register_metafunction(const char* name, mf_sign_func_inout* f) {
+    active_instance->register_one(name, f);
+}
+
+} // namespace meta
+
 auto parser::apply_type_metafunctions( declaration_node& n )
     -> bool
 {
     assert(n.is_type());
 
     //  Get the reflection state ready to pass to the function
-    auto cs = meta::compiler_services{ &errors, &includes, generated_tokens };
-    auto rtype = meta::type_declaration{ &n, cs };
+    auto csb = meta::compiler_services_base{ &libraries, &errors, &includes, generated_tokens, &generated_declarations };
+    auto rtypeimpl = meta::type_declaration_impl{ csb, &n };
+
+    return apply_metafunctions(
+        n,
+        rtypeimpl,
+        [&](std::string const& msg) { error( msg, false ); },
+        [&](
+            std::string_view caller,
+            std::string const& name,
+            meta::type_declaration& decl
+        ) -> bool { return libraries.try_apply(caller, name, decl); }
```

Comment:
Why not directly provide the libraries object?

Maybe because reflect_libraries would then need to be defined in `cpp2reflect_api.h`?

If so, would it make sense to move `apply_metafunction` out of cpp2reflect_api into `reflect_libraries` and just call a function once in `cpp2reflect_api` which registers all the internal metafunctions?

Reply:
Not sure if I follow... The main reason we do this wrapping and pass a lambda instead of passing the object, is to entirely decouple the existence of `reflect_libraries` with the rest of the implementation, it has nothing to do with exposing things in reflect_api or anything like that. Same thing is done for printing errors.

Reply:
Ok, I can understand the decoupling argument.